liblwgeom/lwgeom.c

```diff
@@ -2534,3 +2534,181 @@ void lwgeom_trim_bits_in_place(LWGEOM* geom, int32_t prec_x, int32_t prec_y, int
+			POINT4D *out = lwalloc(sizeof(POINT4D) * lwmline->ngeoms * 2);
+			uint32_t n = 0;
+			uint32_t i, j, k;
+			for (i = 0; i < lwmline->ngeoms; i++)
+			{
+				POINTARRAY *line_points = lwmline->geoms[i]->points;
+
+				for (k = 0; k < 2; k++)
+				{
+					POINT4D pt;
+					getPoint4d_p(line_points, k == 1 ? line_points->npoints - 1 : 0, &pt);
+
+					int seen = LW_FALSE;
+					for (j = 0; j < n; j++)
+					{
+						if (memcmp(&(out[j]), &pt, sizeof(POINT4D)) == 0)
+						{
+							seen = LW_TRUE;
+							break;
+						}
+					}
+
+					if (seen)
+					{
+						memcpy(&(out[j]), &(out[n - 1]), sizeof(POINT4D));
+						n--;
+					}
+					else
+					{
+						memcpy(&(out[n]), &pt, sizeof(POINT4D));
+						n++;
+					}
+				}
+
+			}
+
+			LWMPOINT *lwmpoint = lwmpoint_construct_empty(srid, hasz, hasm);
+			for (i = 0; i < n; i++) {
+				lwmpoint_add_lwpoint(lwmpoint, lwpoint_make(srid, hasz, hasm, &(out[i])));
+			}
+
+			lwfree(out);
+
+			lwresult = (LWGEOM *)lwmpoint;
+
+			break;
+		}
+		case TRIANGLETYPE:
+		{
+			LWTRIANGLE *lwtriangle = lwgeom_as_lwtriangle(lwgeom);
+			POINTARRAY *points = ptarray_clone_deep(lwtriangle->points);
+			lwresult = (LWGEOM *)lwline_construct(srid, gbox_clone(lwtriangle->bbox), points);
+			break;
+		}
+		case POLYGONTYPE:
+		{
+			LWPOLY *lwpoly = lwgeom_as_lwpoly(lwgeom);
+
+			if (lwpoly->nrings == 1)
+			{
+				POINTARRAY *ring = ptarray_clone_deep(lwpoly->rings[0]);
+				lwresult = (LWGEOM *)lwline_construct(srid, 0, ring);
+			}
+			else
+			{
+				LWMLINE *lwmline = lwmline_construct_empty(srid, hasz, hasm);
+				for (uint32_t i = 0; i < lwpoly->nrings; i++)
+				{
+					POINTARRAY *ring = ptarray_clone_deep(lwpoly->rings[i]);
+					lwmline_add_lwline(lwmline, lwline_construct(srid, 0, ring));
+				}
+
+				lwresult = (LWGEOM *)lwmline;
+			}
+
+			break;
+		}
+		case MULTIPOLYGONTYPE:
+		{
+			LWMLINE *lwmline = lwmline_construct_empty(srid, hasz, hasm);
+			LWMPOLY *lwmpoly = lwgeom_as_lwmpoly(lwgeom);
+			for (uint32_t i = 0; i < lwmpoly->ngeoms; i++)
+			{
+				LWPOLY *lwpoly = lwmpoly->geoms[i];
+				for (uint32_t j = 0; j < lwpoly->nrings; j++)
+				{
+					POINTARRAY *ring = ptarray_clone_deep(lwpoly->rings[j]);
+					lwmline_add_lwline(lwmline, lwline_construct(srid, 0, ring));
+				}
+			}
+
+			lwresult = (LWGEOM *)lwmline;
+			break;
+		}
+		case TINTYPE:
+		{
+			LWTIN *lwtin = lwgeom_as_lwtin(lwgeom);
+
+			if (lwtin->ngeoms == 1)
+			{
+				POINTARRAY *points = ptarray_clone_deep(lwtin->geoms[0]->points);
+				lwresult = (LWGEOM *)lwline_construct(srid, lwtin->geoms[0]->bbox, points);
+			}
+			else
+			{
+				LWMLINE *lwmline = lwmline_construct_empty(srid, hasz, hasm);
+
+				for (uint32_t i = 0; i < lwtin->ngeoms; i++)
+				{
+					POINTARRAY *points = ptarray_clone_deep(lwtin->geoms[i]->points);
+					lwmline_add_lwline(lwmline, lwline_construct(srid, lwtin->geoms[i]->bbox, points));
+				}
+
+				lwresult = (LWGEOM *)lwmline;
+			}
+
+			break;
+		}
+		case COLLECTIONTYPE:
```

Comment:
From OGC:
> The boundary of an arbitrary collection of geometric objects whose interiors are disjoint consists of geometric objects drawn from the boundaries of the element geometric objects by application of the “mod 2” union rule (Bibliographic Reference [1], section 3.12.3.2).

So my assumption that boundary of geometry collection is collection of elements boundaries must to be wrong. I want to make it to meet the specification but not sure how boundary is supposed to be calculate for collection whose interiors are _not_ disjoint.

Reply:
From OGC:
> The subtypes of GeometryCollection, to be presented next, shall ensure the following constraints, which are not automatically ensured by aggregates of the Spatial schema. These constraints are summarized as follows.
> a) For every element in a GeometryCollection, its interior shall be disjoint to the interior of every other, but distinct element of the same GeometryCollection.
> b) For every boundary of an element in a GeometryCollection, it may only intersect with a boundary of another, but distinct element at most in a finite number of points.

Does that mean that it is acceptable to output wrong boundary for geometry collection that doesn't meet these constraints?

Reply:
My reading is "If you do UnaryUnion on your collection before, you can expect a boundary to be then edge between meat and exterior of collection. If you don't, you get back something you may not like."